internal/services/v1/watch.go

```diff
@@ -52,11 +52,15 @@ func (ws *watchServer) Watch(req *v1.WatchRequest, stream v1.WatchService_WatchS
 
 	var afterRevision datastore.Revision
 	if req.OptionalStartCursor != nil && req.OptionalStartCursor.Token != "" {
-		decodedRevision, err := zedtoken.DecodeRevision(req.OptionalStartCursor, ds)
+		decodedRevision, tokenStatus, err := zedtoken.DecodeRevision(req.OptionalStartCursor, ds)
 		if err != nil {
 			return status.Errorf(codes.InvalidArgument, "failed to decode start revision: %s", err)
 		}
 
+		if tokenStatus == zedtoken.StatusMismatchedDatastoreID {
```

Comment:
I think context is missing on _why_ we are choosing to fail here. I would recommend adding a comment to clarify, since the implications are not immediately clear.

I'm trying to think what type of scenario are we trying to protect from by failing here, but also potential use-cases we could be hampering by doing so.

- Clearly zedtoken for a different datastore type should fail as those can fail in unexpected ways, even though we could potentially find ways some of them interoperate.
- Same datastore type, but different instances: we can't guarantee to observe the same transactions at the same timestamp in different instances.

In the case of Watch, reusing a zedtoken in different instances could lead to missing events.

The only scenario I can think that could be impacted by this are dual-writing to a new SpiceDB instance and doing an eventual cut-over (e.g. deep schema refactor). API calls with `at_least_as_fresh` can recover by momentarily switching to `minimize_latency` until the tokens are refreshed (although this could take _very long_ as it depends on resource changes! the system could be running in minimize for _a long time_). A cut-over to the Watch API will simply not work, regardless of the changes in this PR, as it can only work if they represent the same state of the world.

Long story short, I think this is the right call, and I wanted to think out lout and leave a trace of the why.

Reply:
Added a comment